oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/ProtectedPropertyTest.java

```diff
@@ -0,0 +1,376 @@
+import javax.jcr.nodetype.NodeTypeTemplate;
+import javax.jcr.nodetype.PropertyDefinitionTemplate;
+
+import org.apache.jackrabbit.oak.fixture.NodeStoreFixture;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test coverage for common JCR operations related to importing content.
+ * <p>
+ * Note that the purpose of these tests is not to check conformance with the JCR
+ * specification, but to observe the actual behavior of the implementation
+ * (which may be hard to change).
+ */
+public class ProtectedPropertyTest extends AbstractRepositoryTest {
+
+    private Session session;
+    private Node testNode;
+    private static String TEST_NODE_NAME = "ImportOperationsTest";
+    private static String TEST_NODE_NAME_REF = "ImportOperationsTest-Reference";
+    private static String TEST_NODE_NAME_TMP = "ImportOperationsTest-Temp";
+
+    public ProtectedPropertyTest(NodeStoreFixture fixture) {
+        super(fixture);
+    }
+
+    @Before
+    public void setup() throws RepositoryException {
+        this.session = getAdminSession();
+        this.testNode = session.getRootNode().addNode("import-tests", NodeType.NT_UNSTRUCTURED);
+        session.save();
+    }
+
+    @Test
+    public void jcrMixinCreatedOnNtUnstructured() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        try {
+            test.setProperty("jcr:created", false);
+            session.save();
+            test.addMixin(NodeType.MIX_CREATED);
+            session.save();
+            assertEquals(false, test.getProperty("jcr:created").getBoolean());
+            // in Oak, existing properties are left as-is (even the property
+            // type), which means that after adding the mixin:created type, the
+            // state of the node might be inconsistent with the mixin:created
+            // type. This may come as a surprise, but is allowed as
+            // implementation specific behavior, see
+            // https://developer.adobe.com/experience-manager/reference-materials/spec/jcr/2.0/3_Repository_Model.html#3.7.11.7%20mix:created
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredBeforeSettingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        try {
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            test.addMixin(NodeType.MIX_REFERENCEABLE);
+            session.save();
+            // JCR spec
+            // (https://developer.adobe.com/experience-manager/reference-materials/spec/jcr/2.0/3_Repository_Model.html#3.8%20Referenceable%20Nodes)
+            // requests an "auto-created" property, so it might be a surprise
+            // that Oak actually keeps the application-assigned previous value.
```

Comment:
The algorithm used to create the property value is free to use the existing one and I think that is also the most sane approach, so is it really that surprising?

Reply:
the spec says "autocreated" :-)